### PR TITLE
CLDR-14232 Change blood glucose measures to be compatible

### PR DESCRIPTION
**Pull request by macchiati**
The goal is to have the two measures of blood-glucose be convertible. That means changing one of them to be the same (resolved) units as the other. This was done, like ofhg, by introducing an internal measure that has the right units and factor to work. 

ldmlSupplemental: added descriptions for unit elements (this was part of another ticket but was useful to pull forward)
Locale files: added ofglucose in concentr-milligram-per-deciliter wherever it was,
units.xml: added conversions between the two measures of blood glucose, with another oddball internal unit ofglucose, merged preferences, added alias for old unit
unit.xml: replaced old by new unit, added internal unit ofglucose
DtdData: same addition
Rational: added a utility
UnitConverter: cleaned up problems
 - some of those where from an exposed preexisting problem with kilogram being the base SI unit
 - a null check on metricUnit instead of metricUnit.value,
 - and a duplication long-short ID mapping
TestUnits: added tests for the change, and for kilogram handling
TestValidity: used better error exception

##### Checklist

- [ ] Issue filed: https://unicode-org.atlassian.net/browse/CLDR-14232
- [ ] Updated PR title and link in previous line to include Issue number

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by macchiati** [comment]
This touches a bunch of places, so I'd like to get it reviewed as soon as feasible.

**Review by sffc — APPROVED** [review]

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java [hunk 1]
```diff
@@ -71,7 +69,7 @@
     private Map<String,String> fixDenormalized;
     private ImmutableMap<String, UnitId> idToUnitId;
 
-    public final BiMap<String,String> SHORT_TO_LONG_ID;
+    public final BiMap<String,String> SHORT_TO_LONG_ID = Units.LONG_TO_SHORT.inverse();
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java:72]: better!

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java [hunk 2]
```diff
@@ -365,15 +363,15 @@ public UnitIterator iterator() {
...
-        SHORT_TO_LONG_ID = ImmutableBiMap.copyOf(_SHORT_TO_LONG_ID);
+//        // we need to pass in the validity so it is for the same CLDR version as the converter
+//        Set<String> VALID_UNITS = validity.getStatusToCodes(LstrType.unit).get(Status.regular);
+//        Map<String,String> _SHORT_TO_LONG_ID = new LinkedHashMap<>();
+//        for (String longUnit : VALID_UNITS) {
+//            int dashPos = longUnit.indexOf('-');
+//            String coreUnit = longUnit.substring(dashPos+1);
+//            _SHORT_TO_LONG_ID.put(coreUnit, longUnit);
+//        }
+//        SHORT_TO_LONG_ID = ImmutableBiMap.copyOf(_SHORT_TO_LONG_ID);
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java:374]:
```suggestion
```

@btangmu has got me on the deletionist bandwagon… this code can't be used anymore due to the .inverse(). could be fixed in a follow up PR.
tom's got

File: tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java [hunk 3]
```diff
@@ -1218,10 +1220,22 @@ public boolean isSimpleBaseUnit(String unit) {
         for (Entry<String, V> entry : unitMap.entrySet()) {
             String prefix = entry.getKey();
             if (unit.startsWith(prefix)) {
+                String result = unit.substring(prefix.length());
+                // We have to do a special hack for kilogram, but only for the Rational case.
+                // The Integer case is used for name construction, so that is ok.
```
Comment (srl295) [inline tools/cldr-code/src/main/java/org/unicode/cldr/util/UnitConverter.java:1225]: is there a spec impact here?

File: tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java [hunk 4]
```diff
@@ -1918,13 +1938,16 @@ public void TestFallbackNames() {
...
+            default: resolvedFile = resolvedFileRaw; break;
             }
 
             String shortUnitId = row[1];
             String pluralCategory = row[2];
             String caseVariant = row[3];
             String expectedName = row[4];
+            if (shortUnitId.equals("gigasecond-milligram")) {
+                int debug = 0;
+            }
```
Comment (srl295) [inline tools/cldr-code/src/test/java/org/unicode/cldr/unittest/TestUnits.java:1950]:
```suggestion
```

**Review by srl295 — APPROVED** [review]
LGTM with comments.  A future exercise would be tests for `getIntPowerOfTen()` etc, verifying code coverage.

**Comment by macchiati** [comment]
I'll do a followup PR with some cleanup. However, I'd like to get this in asap so that it doesn't block Hugo. I appear to be blocked by "@github-actions Ansible Lint / build (pull_request) Failing after 28m — build" "Try the provision". It doesn't have an error message or any indication of how to address the failure.

**Review by hugovdm — COMMENTED** [review]
Looks good to me - thanks!

**Comment by srl295** [comment]
current CLDR + `cldr-staging` v38.1  cannot regenerate JSON, fails on `jv.xml`:

```
[000%]:	 Beginning parallel process of 788 file(s)
[WARNING]
java.lang.IllegalArgumentException: java.lang.IllegalArgumentException: Missing Map Comparator value(s): concentr-milligram-per-deciliter(null),	length-centimeter(91),
    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0 (Native Method)
    at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance (NativeConstructorAccessorImpl.java:64)
    at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance (DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstanceWithCaller (Constructor.java:500)
    at java.lang.reflect.Constructor.newInstance (Constructor.java:481)
    at java.util.concurrent.ForkJoinTask.getThrowableException (ForkJoinTask.java:600)
    at java.util.concurrent.ForkJoinTask.reportException (ForkJoinTask.java:678)
    at java.util.concurrent.ForkJoinTask.invoke (ForkJoinTask.java:737)
    at java.util.stream.Nodes.collect (Nodes.java:336)
    at java.util.stream.ReferencePipeline.evaluateToNode (ReferencePipeline.java:109)
    at java.util.stream.AbstractPipeline.evaluate (AbstractPipeline.java:545)
    at java.util.stream.AbstractPipeline.evaluateToArrayNode (AbstractPipeline.java:260)
    at java.util.stream.ReferencePipeline.toArray (ReferencePipeline.java:517)
    at java.util.stream.ReferencePipeline.toArray (ReferencePipeline.java:523)
    at org.unicode.cldr.json.Ldml2JsonConverter.processDirectory (Ldml2JsonConverter.java:1445)
    at org.unicode.cldr.json.Ldml2JsonConverter.main (Ldml2JsonConverter.java:154)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:832)
Caused by: java.lang.IllegalArgumentException: Missing Map Comparator value(s): concentr-milligram-per-deciliter(null),	length-centimeter(91),
    at org.unicode.cldr.util.MapComparator.compare (MapComparator.java:147)
    at org.unicode.cldr.util.DtdData$DtdComparator.xpathComparator (DtdData.java:881)
    at org.unicode.cldr.util.DtdData$DtdComparator.compare (DtdData.java:809)
    at org.unicode.cldr.util.DtdData$DtdComparator.compare (DtdData.java:804)
    at java.util.TreeMap.put (TreeMap.java:787)
    at java.util.TreeMap.put (TreeMap.java:534)
    at java.util.TreeSet.add (TreeSet.java:255)
    at java.util.Iterator.forEachRemaining (Iterator.java:133)
    at java.util.Collections$UnmodifiableCollection$1.forEachRemaining (Collections.java:1056)
    at org.unicode.cldr.util.CLDRFile.iterator (CLDRFile.java:1244)
    at org.unicode.cldr.json.Ldml2JsonConverter.mapPathsToSections (Ldml2JsonConverter.java:365)
    at org.unicode.cldr.json.Ldml2JsonConverter.lambda$processDirectory$10 (Ldml2JsonConverter.java:1432)
    at java.util.stream.ReferencePipeline$3$1.accept (ReferencePipeline.java:195)
    at java.util.HashMap$KeySpliterator.forEachRemaining (HashMap.java:1694)
    at java.util.stream.AbstractPipeline.copyInto (AbstractPipeline.java:484)
    at java.util.stream.AbstractPipeline.wrapAndCopyInto (AbstractPipeline.java:474)
    at java.util.stream.Nodes$CollectorTask.doLeaf (Nodes.java:2191)
    at java.util.stream.Nodes$CollectorTask.doLeaf (Nodes.java:2157)
    at java.util.stream.AbstractTask.compute (AbstractTask.java:327)
    at java.util.concurrent.CountedCompleter.exec (CountedCompleter.java:746)
    at java.util.concurrent.ForkJoinTask.doExec (ForkJoinTask.java:290)
    at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec (ForkJoinPool.java:1016)
    at java.util.concurrent.ForkJoinPool.scan (ForkJoinPool.java:1665)
    at java.util.concurrent.ForkJoinPool.runWorker (ForkJoinPool.java:1598)
    at java.util.concurrent.ForkJoinWorkerThread.run (ForkJoinWorkerThread.java:183)
```


The rough command line is:
```shell
$ mvn --file=../cldr/tools/pom.xml -DCLDR_DIR=../cldr-staging/production exec:java -pl cldr-code '-Dexec.args=-m .* -p true -o true -r true -t main -d ./cldr-json -s contributed -V 38.1.0'
```

the quick repro in code will be something like this:

```java
CLDRFile jv = … // jv.xml
        for (Iterator<String> it = file.iterator("", DtdData.getInstance(DtdType.ldml).getDtdComparator(null)); it.hasNext();) {
   // noop is fine
}
```

**Comment by srl295** [comment]
https://github.com/unicode-org/cldr/pull/983 attempts to get around the above by partially reverting the DtdData part..

**Comment by hugovdm** [comment]
Tested on ICU: I believe we need an update to common/testData/units/unitPreferencesTest.txt as well: there's a test case that I believe depends on the removed mass-density blood-glucose usage. (The alias added impacts the expected unit, but not the input unit, and thus still gets the unit test classified as mass-density, not concentration.)

**Comment by macchiati** [comment]
Good catch. I can regenerate that.

Mark

**Comment by younies** [comment]
@macchiati , we test units preferences but we do not test formatting. we need to test formatting for aliases too.

**Pull request merged**